compiler/src/dotty/tools/dotc/core/Definitions.scala

```diff
@@ -336,11 +344,29 @@ class Definitions {
     ScalaPackageClass, tpnme.Nothing, AbstractFinal, List(AnyClass.typeRef))
   def NothingType: TypeRef = NothingClass.typeRef
   @tu lazy val RuntimeNothingModuleRef: TermRef = ctx.requiredModuleRef("scala.runtime.Nothing")
-  @tu lazy val NullClass: ClassSymbol = enterCompleteClassSymbol(
-    ScalaPackageClass, tpnme.Null, AbstractFinal, List(ObjectClass.typeRef))
+  @tu lazy val NullClass: ClassSymbol = {
+    val parent = if (ctx.explicitNulls) AnyType else ObjectType
```

Comment:
An issue with Having `Null` under `AnyVal` is that `NotNull` is harder to express. So, maybe hold off with that for now.

Reply:
Hum, what `NotNull` are you referring to? With `Null` under `AnyVal`, a type `T` is guaranteed not to contain `null` if `T <: AnyRef`. It's as simple as that.

Reply:
The type could also be a primitive type like Int or Boolean, or a value class. `NotNull` is needed for type signatures like a Java-like map get:
```
def [K, V <: NotNull](m: Map[K, V]) get (k: K): V | Null
```

Reply:
Ah, I see.

Well TBH I'm not sure that using this kind of signature is something we would want to encourage. We want people *not* to use `Null` in a normal Scala program.

Anyway, you can still define `NotNull` as
```scala
type NotNull = AnyRef | Boolean | Char | Byte | Short | Int | Long | Float | Double | Unit
```
Yes, that would take longer to typecheck against, but I would certainly hope that usage of `NotNull` in a regular Scala program would be very limited, and used only when performance is so important that you're willing to deal with `Null`. And if you're concerned about performance, then probably you're concerned about boxing and that means you probably don't want to accept other primitive types, so `<: AnyRef` is what you would use.

(Also, yes, my `NotNull` does not admit custom `AnyVal` classes, but for the same reasons I don't think this is a likely use case.)

Reply:
Presumably `AnyVal` classes will be phased out in favour of opaque types?

Reply:
A related issue we've encountered in the Dotty community build is uses of Option.apply. We'd like it to have a signature like
`apply[A](x: A|Null): Option[A]`
but here we want to ensure that the `A` is not null.

Reply:
The case of `Option.apply` is not a compelling example, for two reasons.

First, it is not required for type-safety of `Option.apply` that `A` be guaranteed non-null. It is perfectly valid to call
```scala
def f[A](x: A, default: A): A = {
  val y = Option[A](x)
  y.getOrElse(default)
}
f[String | Null](null, "default") // returns "default"
```
This function makes no sense, but my point is that I have not broken type-safety by calling `Option.apply` with an `A` that is nullable.

If you do something a bit-nonsensical with your `Option.apply`, well you'll just get a result type that is not as good as you might hope for, but that will only cause downstream type errors; it won't cause run-time exceptions. So this is perfectly fine.

The other reason is that `Option.apply` is a very misleading API to begin with, one that is already largely misused in the wild. How many times have I seen people call `Option(x)` when they really wanted `Some(x)` but didn't know it. This should not be blamed on users, but on the API: the general contract of `Companion(x)` is to return the *unit* of of the type `Companion` for the value `x`. For example, `List(x)` always returns a list of 1 element containing `x`. `Option(x)` violates that general contract by treating `null` in a non-type-parametric way. `Option.apply` should never have been called `apply`; it should have been called `fromNullable`. A correct `Option.apply` should return `Some(null)` when given `null`, just like `List(null)` returns `null :: Nil`. So basically using `Option.apply` as a motivation for a language design decision is extremely misguided, because `Option.apply` is flawed to begin with.

Reply:
> `Option(x)` violates that general contract by treating `null` in a non-type-parametric way

Is `Try.apply` equally flawed?

Reply:
No, why would it?

Reply:
Maybe, but in a slightly different way.

Reply:
could that be fixed with:
```scala
def apply[A <: NotNull](a: A | Null): Option[A] = ..
```
then your example would be:
```scala
def f[A <: NotNull](x: A, default: A): A = {
  val y = Option[A](x)
  y.getOrElse(default)
}
f[String | Null](null, "default") // fails compilation: `String | Null` is not `NotNull`
```

Reply:
The motivating example is code like:
```Scala
Option(somePossiblyNullThing) match {
  case Some(s) => ???
  case None => ???
  case null => ???
}
```

It would be nice to not have to include the last case.

More concretely, in minitest, there is the line:
```Scala
Option(source.getMessage).filterNot(_.isEmpty)
```
This fails to compile because the `Option()` returns an `Option[T|Null]`, so then the `_` could be null, so you can't call `.isEmpty` on it. To get it to compile, we currently have to rewrite the above to:
```Scala
Option(source.getMessage).filterNot(_.nn.isEmpty)
```
I think this defeats the main purpose of `Option.apply`, which is to get rid of the possibility of `null` by changing it to `None`.

Reply:
That seems to be an inference issue, not a subtyping/type hierarchy issue. We could teach type inference that when it tries to infer `A` to unify `A | Null` with some `T | Null`, it should choose `A = T` instead of `A = T | Null`.